central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -3,18 +3,82 @@ package datastore
 import (
 	"context"
 
+	"github.com/heimdalr/dag"
 	"github.com/stackrox/rox/central/resourcecollection/datastore/index"
 	"github.com/stackrox/rox/central/resourcecollection/datastore/search"
-	"github.com/stackrox/rox/central/resourcecollection/datastore/store/postgres"
+	"github.com/stackrox/rox/central/resourcecollection/datastore/store"
 	v1 "github.com/stackrox/rox/generated/api/v1"
 	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/logging"
 	pkgSearch "github.com/stackrox/rox/pkg/search"
 )
 
+var (
+	log = logging.LoggerForModule()
+)
+
 type datastoreImpl struct {
-	storage  postgres.Store
+	storage  store.Store
 	indexer  index.Indexer
 	searcher search.Searcher
+
+	graph *dag.DAG
+}
+
+func (ds *datastoreImpl) initGraph(ctx context.Context) error {
+	if ds.graph != nil {
+		return nil
+	}
+	ds.graph = dag.NewDAG()
+
+	// first walk adds vertices
+	err := ds.storage.Walk(ctx, func(obj *storage.ResourceCollection) error {
+		return ds.graph.AddVertexByID(obj.GetId(), obj.GetId())
+	})
+	if err != nil {
+		return err
+	}
+
+	// second walk adds edges
+	return ds.storage.Walk(ctx, func(obj *storage.ResourceCollection) error {
+		for _, edge := range obj.GetEmbeddedCollections() {
+			err = ds.graph.AddEdge(obj.GetId(), edge.GetId())
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+func (ds *datastoreImpl) addCollectionToGraph(ctx context.Context, obj *storage.ResourceCollection) error {
+	if err := ds.initGraph(ctx); err != nil {
+		return err
+	}
+
+	err := ds.graph.AddVertexByID(obj.GetId(), obj.GetId())
+	if err != nil {
+		return err
+	}
+	for _, edge := range obj.GetEmbeddedCollections() {
+		err = ds.graph.AddEdge(obj.GetId(), edge.GetId())
+		if err != nil {
+			deleteErr := ds.graph.DeleteVertex(obj.GetId())
+			if deleteErr != nil {
+				log.Errorf("Failed to delete collection, might result in bad state (%v)", deleteErr)
```

Comment:
This is more of a warning, no? It's not actually going to prevent anything and you aren't deleting the collection from the datastore (so the message is incorrect as well). All it does is prevent finding cycles. A restart of the pod will fix it too.

Reply:
If the vertex doesn't get deleted any following collection with the same id would not be able to be added. I think this line is more of a precaution, "proper error handling" and whatnot. I think logically the line would never be run, but if it did it would be really nice to know. According to the docs the function returns error if the id is empty (something that would cause the `AddVertexById` function call to return error earlier in the function), or unknown (something that should never happen because we know we just added the id to the graph)

Reply:
Fair enough to keep it as an error, but I think the error message is misleading to the customer who might be confused about why it mentions deleting a collection when they are trying to add.

Perhaps reworded to "Failed to remove collection <id> from internal state object" or something like that